// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: contacts.proto

package model

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Contact with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Contact) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Contact with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in ContactMultiError, or nil if none found.
func (m *Contact) ValidateAll() error {
	return m.validate(true)
}

func (m *Contact) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for WaId

	// no validation rules for Input

	// no validation rules for Status

	if all {
		switch v := interface{}(m.GetProfile()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ContactValidationError{
					field:  "Profile",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ContactValidationError{
					field:  "Profile",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProfile()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ContactValidationError{
				field:  "Profile",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ContactMultiError(errors)
	}
	return nil
}

// ContactMultiError is an error wrapping multiple validation errors returned
// by Contact.ValidateAll() if the designated constraints aren't met.
type ContactMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactMultiError) AllErrors() []error { return m }

// ContactValidationError is the validation error returned by Contact.Validate
// if the designated constraints aren't met.
type ContactValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactValidationError) ErrorName() string { return "ContactValidationError" }

// Error satisfies the builtin error interface
func (e ContactValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContact.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactValidationError{}

// Validate checks the field values on ContactRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ContactRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ContactRequestMultiError,
// or nil if none found.
func (m *ContactRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Blocking

	// no validation rules for ForceCheck

	if len(errors) > 0 {
		return ContactRequestMultiError(errors)
	}
	return nil
}

// ContactRequestMultiError is an error wrapping multiple validation errors
// returned by ContactRequest.ValidateAll() if the designated constraints
// aren't met.
type ContactRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactRequestMultiError) AllErrors() []error { return m }

// ContactRequestValidationError is the validation error returned by
// ContactRequest.Validate if the designated constraints aren't met.
type ContactRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactRequestValidationError) ErrorName() string { return "ContactRequestValidationError" }

// Error satisfies the builtin error interface
func (e ContactRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactRequestValidationError{}

// Validate checks the field values on ContactResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ContactResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ContactResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ContactResponseMultiError, or nil if none found.
func (m *ContactResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ContactResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetContacts() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ContactResponseValidationError{
						field:  fmt.Sprintf("Contacts[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ContactResponseValidationError{
						field:  fmt.Sprintf("Contacts[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactResponseValidationError{
					field:  fmt.Sprintf("Contacts[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ContactResponseMultiError(errors)
	}
	return nil
}

// ContactResponseMultiError is an error wrapping multiple validation errors
// returned by ContactResponse.ValidateAll() if the designated constraints
// aren't met.
type ContactResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ContactResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ContactResponseMultiError) AllErrors() []error { return m }

// ContactResponseValidationError is the validation error returned by
// ContactResponse.Validate if the designated constraints aren't met.
type ContactResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactResponseValidationError) ErrorName() string { return "ContactResponseValidationError" }

// Error satisfies the builtin error interface
func (e ContactResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactResponseValidationError{}

// Validate checks the field values on Contact_Profile with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *Contact_Profile) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Contact_Profile with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// Contact_ProfileMultiError, or nil if none found.
func (m *Contact_Profile) ValidateAll() error {
	return m.validate(true)
}

func (m *Contact_Profile) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if len(errors) > 0 {
		return Contact_ProfileMultiError(errors)
	}
	return nil
}

// Contact_ProfileMultiError is an error wrapping multiple validation errors
// returned by Contact_Profile.ValidateAll() if the designated constraints
// aren't met.
type Contact_ProfileMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m Contact_ProfileMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m Contact_ProfileMultiError) AllErrors() []error { return m }

// Contact_ProfileValidationError is the validation error returned by
// Contact_Profile.Validate if the designated constraints aren't met.
type Contact_ProfileValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e Contact_ProfileValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e Contact_ProfileValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e Contact_ProfileValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e Contact_ProfileValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e Contact_ProfileValidationError) ErrorName() string { return "Contact_ProfileValidationError" }

// Error satisfies the builtin error interface
func (e Contact_ProfileValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContact_Profile.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = Contact_ProfileValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = Contact_ProfileValidationError{}
